module/Photo/src/Photo/Mapper/Photo.php

```diff
@@ -30,12 +30,79 @@ public function __construct(EntityManager $em)
     }
 
     /**
+     * Returns the next photo in the album to display
+     * 
+     * @param \Photo\Model\Photo $photo 
+     */
+    public function getNextPhoto($photo)
+    {
+        $qb = $this->em->createQueryBuilder();
+
+        $qb->select('a')
+                ->from('Photo\Model\Photo', 'a')
+                ->where('a.id > ?1 AND a.album = ?2');
+        $qb->setParameter(1, $photo->getId());
+        $qb->setParameter(2, $photo->getAlbum());
+        $qb->addOrderBy('a.id', 'ASC');
+        $qb->setMaxResults(1);
+        $res = $qb->getQuery()->getResult();
+        return empty($res) ? null : $res[0];
+    }
+
+    /**
+     * Returns the previous photo in the album to display
+     * 
+     * @param \Photo\Model\Photo $photo 
+     */
+    public function getPreviousPhoto($photo)
+    {
+        $qb = $this->em->createQueryBuilder();
+
+        $qb->select('a')
+                ->from('Photo\Model\Photo', 'a')
+                ->where('a.id < ?1 AND a.album = ?2');
+        $qb->setParameter(1, $photo->getId());
+        $qb->setParameter(2, $photo->getAlbum());
+        $qb->addOrderBy('a.id', 'DESC');
+        $qb->setMaxResults(1);
+        $res = $qb->getQuery()->getResult();
+        return empty($res) ? null : $res[0];
+    }
+
+    /**
+     * Returns a random photo from the specified album
+     * 
+     * @param Photo\Model\Album $album
+     * @return Photo\Model\Photo
+     */
+    public function getRandomPhoto($album)
+    {
+        $qb = $this->em->createQueryBuilder();
+
+        /**
+         * This is not exactly the right way to do this. There is no way to just
+         * get a random row, unless we use a raw sql query.
+         * The doctrine team refuses to implement any 
+         * random functionality. (http://www.doctrine-project.org/jira/browse/DDC-950)
+         * 
+         */
+        $qb->select('a')
+                ->from('Photo\Model\Photo', 'a')
+                ->where('a.album = ?1');
+        $qb->setParameter(1, $album);
+        $qb->setMaxResults(20);
+        $res = $qb->getQuery()->getResult();
+        return empty($res) ? null : $res[array_rand($res)];
```

Comment:
Dat is inderdaad lastig. Ik zie hier ook niet zo snel een nette oplossing. Ik ga er iniedergeval over nadenken.

Overigens, is er zoiezo geen snelle methode hiervoor. Ook in SQL zelf is zoiets eigenlijk ook inefficient geimplementeerd (hiervoor worden eerst alle mogelijke resultaten opgehaald, waarna er random gesorteerd wordt, waarna het eerste resultaat teruggegeven wordt).

Als je een goed idee hebt dat qua UI ook werkt, maar dus niet random is, mag het natuurlijk ook.